README.md
Node Airbrake
=============

[![Circle CI](https://circleci.com/gh/airbrake/node-airbrake.svg?style=shield)](https://circleci.com/gh/airbrake/node-airbrake)
[![npm version](https://badge.fury.io/js/airbrake.svg)](https://badge.fury.io/js/airbrake)
[![Documentation Status](http://inch-ci.org/github/airbrake/node-airbrake.svg?branch=master)](http://inch-ci.org/github/airbrake/node-airbrake)
[![Downloads](https://img.shields.io/npm/dt/airbrake.svg)](https://www.npmjs.com/package/airbrake)

![Node Airbrake][arthur-node]

* [Node Airbrake README][node-airbrake]

Introduction
------------

_Node Airbrake_ is a Node.js notifier for [Airbrake][airbrake-io], the leading
exception reporting service. The library provides minimalist API that enables
the ability to send _any_ Node exception to the Airbrake dashboard. Node
Airbrake provides out-of-box integration with the Express web framework.

Key features
------------

![The Airbrake Dashboard][dashboard]

* Send chosen environment variables (whitelist or blacklist)
* Detect and fix circular references in error context information
* Support for all features of the [2.1 notification API][2.1api]
* Support for [long-stack-traces][]
* Optional auto-handler for `uncaughtException` events
* Provides notification URL linking to Airbrake in `notify()` callback
* Timeout Airbrake requests after 30 seconds, you never know

Installation
------------

### NPM

Add the Node Airbrake package to your `package.json`:

```js
{
  "dependencies": {
    "airbrake": "~1.0.3"
  }
}
```

### Manual

Invoke the following command from your terminal:

```sh
npm install airbrake
```

Examples
--------

### Basic example

This is the minimal example that you can use to test Node Airbrake with your
project. The common use case for this module is to catch all `'uncaughtException'`
events on the `process` object and send them to Airbrake:

```js
var airbrake = require('airbrake').createClient(
  '105138', // Project ID
  'fd04e13d806a90f96614ad8e529b2822' // Project key
);
airbrake.handleExceptions();

throw new Error('I am an uncaught exception');
```

Note: the above will re-throw the exception after it has been successfully
delivered to Airbrake, causing your process to exit with status 1.

This can optionally be disabled by passing false to `handleExceptions` (not
recommended):

```js
airbrake.handleExceptions(false);
```

### Filtering errors

There may be some errors thrown in your application that you're not interested in sending to Airbrake, such as errors thrown by 3rd-party libraries.

The Airbrake notifier makes it simple to ignore this chaff while still processing legitimate errors. Add filters to the notifier by providing filter functions to `addFilter`.

`addFilter` accepts the entire [error notice](https://airbrake.io/docs/#create-notice-v3) to be sent to Airbrake, and provides access to the `context`, `environment`, `params`, and `session` values submitted with the notice, as well as the single-element `errors` array with its `backtrace` element and associated backtrace lines.

The return value of the filter function determines whether or not the error notice will be submitted.
  * If null value is returned, the notice is ignored.
  * Otherwise returned notice will be submitted.

An error notice must pass all provided filters to be submitted.

In the following example all errors triggered by admins will be ignored:

```js
airbrake.addFilter(function(notice) {
  if (notice.sessions.admin) {
    // Ignore errors from admin sessions.
    return null;
  }
  return notice;
});
```

Filters can be also used to modify notice payload, e.g. to set environment and application version:

```js
airbrake.addFilter(function(notice) {
  notice.context.environment = 'production';
  notice.context.version = '1.2.3';
  return notice;
});
```

### Manual error delivery

If you want more control over the delivery of your errors, you can also
manually submit errors to Airbrake:

```js
var airbrake = require('airbrake').createClient("your project ID", "your api key");
var err = new Error('Something went terribly wrong');
airbrake.notify(err, function(err, url) {
  if (err) throw err;

  // Error has been delivered, url links to the error in airbrake
});
```

By default only the errors from the production environment will get reported,
so make sure to put `production` in your `NODE_ENV`.

### Express integration

The library provides out-of-box integration with the Express framework. It
supports even old Express versions (starting from `2.x`). Select your version
below and configure accordingly.

#### Express 4.x

```js
var airbrake = require('airbrake').createClient("your project ID", "your api key");
app.use(airbrake.expressHandler());
```

#### Express 3.x

```js
var airbrake = require('airbrake').createClient("your project ID", "your api key");
app.use(app.router);
app.use(airbrake.expressHandler());
```

#### Express 2.x

```js
var airbrake = require('airbrake').createClient("your project ID", "your api key");
app.error(airbrake.expressHandler());
```

API
---

## Adding context to errors

The `notify()` method automatically adds the following context information to
each delivered error:

* **error.class:** (`err.type` string if set, or `'Error'`)
* **error.message:** (`err.message` string)
* **error.backtrace:** (`err.stack` as parsed by [stack-trace][])
* **request.url:** (`err.url`, see `airbrake.url`);
* **request.component:** (`err.component` string if set);
* **request.action:** (`err.action` string if set);
* **request.cgi-data:** (`process.env`, merged all other properties of `err`)
* **request.params:** (`err.params` object if set)
* **request.session:** (`err.session` object if set)
* **server-environment.project-root:** (`airbrake.projectRoot` string if set)
* **server-environment.environment-name:** (`airbrake.env` string)
* **server-environment.app-version:** (`airbrake.appVersion string if set)

You can add additional context information by modifying the error properties
listed above:

``` javascript
var airbrake = require('airbrake').createClient("your project ID", "your api key");
var http = require('http');

http.createServer(function(req, res) {
  if (req.headers['X-Secret'] !== 'my secret') {
    var err = new Error('403 - Permission denied');
    req.writeHead(403);
    req.end(err.message);

    err.url = req.url;
    err.params = {ip: req.socket.remoteAddress};
    airbrake.notify(err);
  }
});
```

Unfortunately `uncaughtException` events cannot be traced back to particular
requests, so you should still try to handle errors where they occur.

[stack-trace]: https://github.com/felixge/node-stack-trace

## Tracking deployments

This client supports Airbrake's [deployment tracking][]:

``` javascript
var airbrake = require('airbrake').createClient("your project ID", "your api key");
var deployment = {
  rev: '98103a8fa850d5eaf3666e419d8a0a93e535b1b2',
  repo: 'git@github.com:felixge/node-airbrake.git',
};

airbrake.trackDeployment(deployment, function(err, params) {
  if (err) {
    throw err;
  }

  console.log('Tracked deployment of %s to %s', params.rev, params.env);
});
```

Check out the `airbrake.trackDeployment()` API docs below for a list of all
options.

Configuration
-------------

### var airbrake = Airbrake.createClient(projectId, key, [env])

`Airbrake.createClient()` returns a new Airbrake instance.

Options
* `projectId` - Your application's Airbrake project ID.
* `key` - Your application's Airbrake API key.
* `env` - The name of the server environment this is running in.

### airbrake.projectId = null

Your application's Airbrake project ID.

### airbrake.key = null

Your application's Airbrake API key.

### airbrake.env = process.env.NODE_ENV || 'development'

The name of the server environment this is running in.

### airbrake.host = 'https://' + os.hostname()

The base url for errors. If `err.url` is not set, `airbrake.host` is used
instead. If `err.url` is a relative url starting with `'/'`, it is appended
to `airbrake.host`. If `err.url` is an absolute url, `airbrake.host` is ignored.

### airbrake.projectRoot = process.cwd()

The root directory of this project.

### airbrake.appVersion = null

The version of this app. Set to a semantic version number, or leave unset.

### airbrake.protocol = 'https'

The protocol to use.

### airbrake.developmentEnvironments = []

Do not post to Airbrake when running in these environments.

### airbrake.timeout = 30 * 1000

The timeout after which to give up trying to notify Airbrake in ms.

### airbrake.proxy = null

The HTTP/HTTPS proxy to use when making requests.

### airbrake.requestOptions = {}

Additional request options that are merged with the default set of options that are passed to `request` during `notify()` and `trackDeployment()`.

### airbrake.whiteListKeys = []

Names of environment variables to send.

### airbrake.blackListKeys = []

Names of environment variables to filter out.

### airbrake.handleExceptions()

Registers a `process.on('uncaughtException')` listener. When an uncaught
exception occurs, the error is sent to Airbrake, and then re-thrown to
kill the process.

### airbrake.expressHandler(disableUncaughtException)

A custom error handler that is used with Express. Integrate with Express
middleware using `app.use()`.

Options:
* `disableUncaughtException`: Disables re-throwing and killing process on uncaught exception.

### airbrake.notify(err, [cb])

Sends the given `err` to airbrake.

The callback parameter receives two arguments, `err, url`. `err` is set if
the delivery to Airbrake failed.

If no `cb` is given, and the delivery fails, an `error` event is emitted. If
there is no listener for this event, node will kill the process as well. This
is done to avoid silent error delivery failure.

### airbrake.trackDeployment([params, [cb]])

Notifies Airbrake about a deployment. `params` is an object with the following
options:

* `env:` The environment being deployed, defaults to `airbrake.env`.
* `user:` The user doing the deployment, defaults to `process.env.USER`.
* `repo:` The github url of this repository. Defaults to `git config --get remote.origin.url`.
* `rev:` The revision of this deployment. Defaults to `git rev-parse HEAD`.

License
-------

The library was originally created by [Felix Geisendörfer](https://github.com/felixge).
The project uses the MIT License. See LICENSE for details.

[arthur-node]: http://s3.amazonaws.com/airbrake-github-assets/node-airbrake/arthur-node.jpeg
[node-airbrake]: https://github.com/airbrake/node-airbrake
[airbrake-io]: https://airbrake.io
[long-stack-traces]: https://github.com/tlrobinson/long-stack-traces
[2.1api]: http://help.airbrake.io/kb/api-2/notifier-api-version-21
[screenshot]: https://github.com/airbrake/node-airbrake/raw/master/screenshot.png
[deployment tracking]: http://help.airbrake.io/kb/api-2/deploy-tracking
[dashboard]: http://s3.amazonaws.com/airbrake-github-assets/node-airbrake/airbrake-dashboard.png


CHANGELOG.md
# Airbrake Changelog

### master

### [v0.3.9][v0.3.9] (February 15, 2016)

* start sending user agent info from Express handler ([#79][79])
* make error handler's killing of process customizable ([#45][45])
* return a Buffer object from `notifyXML` to fix Errbit incompatibility ([#51][51])
* start safely serializing values with `json-stringify-safe` ([#60][60])
* remove hostname from the docs ([#61][61])
* update test suite to use Express 4.x ([#76][76])
* start using Circle CI ([#78][78])
* allow items to be excluded from CGI data ([#65][65])
* documentation fixes ([#68][68] and [#69][69])
* add ability to ignore exceptions ([#74][74])
* update default protocol to https ([#71][71])
* remove uncaught exception from tests ([#72][72])
* refactor `notify` into `_sendRequest` ([#73][73])

[79]:https://github.com/airbrake/node-airbrake/pulls/79
[45]:https://github.com/airbrake/node-airbrake/pull/45
[51]:https://github.com/airbrake/node-airbrake/pull/51
[60]:https://github.com/airbrake/node-airbrake/pull/60
[61]:https://github.com/airbrake/node-airbrake/pull/61
[76]:https://github.com/airbrake/node-airbrake/pull/76
[78]:https://github.com/airbrake/node-airbrake/pull/78
[65]:https://github.com/airbrake/node-airbrake/pull/65
[68]:https://github.com/airbrake/node-airbrake/pull/68
[69]:https://github.com/airbrake/node-airbrake/pull/69
[74]:https://github.com/airbrake/node-airbrake/pull/74
[71]:https://github.com/airbrake/node-airbrake/pull/71
[72]:https://github.com/airbrake/node-airbrake/pull/72
[73]:https://github.com/airbrake/node-airbrake/pull/73
[v0.3.9]: https://github.com/airbrake/node-airbrake/releases/tag/v0.3.9


test/fast/test-request-vars.js
var common = require('../common');
var airbrake = require(common.dir.root).createClient()
var assert = require('assert');
var xmlbuilder = require('xmlbuilder');
var os = require('os');

(function testSettingCustomExclusions() {
  var err = new Error();
  err.domain = { id: 'identifier' };

  airbrake.exclude.push('domain');

  var cgiData = airbrake.cgiDataVars(err);
  assert(!cgiData['err.domain']);
})();

(function testCgiDataFromProcessEnv() {
  var err = new Error();
  var cgiData = airbrake.cgiDataVars(err);

  assert.equal(cgiData['process.pid'], process.pid);
  assert.equal(cgiData['process.uid'], process.getuid());
  assert.equal(cgiData['process.gid'], process.getgid());
  assert.equal(cgiData['process.cwd'], process.cwd());
  assert.equal(cgiData['process.execPath'], process.execPath);
  assert.equal(cgiData['process.version'], process.version);
  assert.equal(cgiData['process.argv'], process.argv);
  assert.ok(cgiData['process.memoryUsage'].rss);
  assert.equal(cgiData['os.loadavg'].length, 3);
  assert.equal(typeof cgiData['os.uptime'], 'number');
  delete cgiData['process.pid'];
  delete cgiData['process.uid'];
  delete cgiData['process.gid'];
  delete cgiData['process.cwd'];
  delete cgiData['process.execPath'];
  delete cgiData['process.version'];
  delete cgiData['process.memoryUsage'];
  delete cgiData['process.argv'];
  delete cgiData['os.loadavg'];
  delete cgiData['os.uptime'];
})();

(function testCustomErrorProperties() {
  var err = new Error();
  err.myKey = 'some value';

  var cgiData = airbrake.cgiDataVars(err);
  assert.equal(cgiData['err.myKey'], err.myKey);
})();

(function testSessionVars() {
  var err = new Error();
  err.session = {foo: 'bar'};

  var session = airbrake.sessionVars(err);
  assert.deepEqual(session, err.session);
})();

(function testParamsVars() {
  var err = new Error();
  err.params = {foo: 'bar'};

  var params = airbrake.paramsVars(err);
  assert.deepEqual(params, err.params);
})();

(function testCircularVars() {
  var vars = {foo: 'bar', circular: {}};
  vars.circular.self = vars.circular;

  // test that no exception is thrown
  var request = xmlbuilder.create().begin('request');
  airbrake.addRequestVars(request, 'params', vars);
})();

(function testAppendErrorXmlWithBadStack() {
  var notice = xmlbuilder.create().begin('notice');
  var err = new Error('oh oh');

  err.stack += '\n    at Array.0 (native)';
  airbrake.appendErrorXml(notice, err);
})();

(function testEmptyErrorMessageDoesNotProduceInvalidXml() {
  // see: https://github.com/felixge/node-airbrake/issues/15
  var err = new Error();
  var xml = airbrake.notifyXml(err, true);

  assert.ok(!/<\/>/.test(xml));
})();


lib/airbrake.js
var HTTP_STATUS_CODES = require('http').STATUS_CODES;

var fs = require('fs');
var os = require('os');
var EventEmitter = require('events').EventEmitter;
var util = require('util');
var request = require('request');
var xmlbuilder = require('xmlbuilder');
var stackTrace = require('stack-trace');
var hashish = require('hashish');
var querystring = require('querystring');

module.exports = Airbrake;
util.inherits(Airbrake, EventEmitter);

function Airbrake() {
  this.key = null;

  this.host = 'https://' + os.hostname();
  this.env = process.env.NODE_ENV || 'development';
  this.projectRoot = null;
  this.appVersion = null;
  this.timeout = 30 * 1000;
  this.developmentEnvironments = ['development', 'test'];
  this.consoleLogError = false;

  this.protocol = 'https';
  this.serviceHost =  process.env.AIRBRAKE_SERVER || 'api.airbrake.io';
  this.requestOptions = {};
  this.ignoredExceptions = [];
}

Airbrake.PACKAGE = (function() {
  var json = fs.readFileSync(__dirname + '/../package.json', 'utf8');
  return JSON.parse(json);
})();

Airbrake.createClient = function(key, env) {
  var instance = new this();
  instance.key = key;
  instance.env = env || instance.env;
  return instance;
};

Airbrake.prototype.expressHandler = function(disableUncaughtException) {
  var self = this;

  if(!disableUncaughtException) {
    process.on('uncaughtException', function(err) {
      self._onError(err, true);
    });
  }

  return function errorHandler(err, req, res, next) {
    if (res.statusCode < 400) res.statusCode = 500;

    err.url = req.url;
    err.component = req.url;
    err.action = req.method;
    err.params = req.body;
    err.session = req.session;

    self._onError(err, false);
    next(err);
  };
};

Airbrake.prototype._onError = function(err, die) {
  var self = this;
  if (!(err instanceof Error)) {
    err = new Error(err);
  }
  self.log('Airbrake: Uncaught exception, sending notification for:');
  self.log(err.stack || err);

  self.notify(err, function(notifyErr, url, devMode) {
    if (notifyErr) {
      self.log('Airbrake: Could not notify service.');
      self.log(notifyErr.stack);
    } else if (devMode) {
      self.log('Airbrake: Dev mode, did not send.');
    } else {
      self.log('Airbrake: Notified service: ' + url);
    }

    if (die) {
      process.exit(1);
    }
  });
};

Airbrake.prototype.handleExceptions = function() {
  var self = this;
  process.on('uncaughtException', function(err) {
    self._onError(err, true);
  });
};

Airbrake.prototype.log = function(str) {
  if(this.consoleLogError) console.error(str);
};

Airbrake.prototype._sendRequest = function(err, cb) {
  var callback = this._callback(cb);

  var body = this.notifyXml(err);

  var options = hashish.merge({
    method: 'POST',
    url: this.url('/notifier_api/v2/notices'),
    body: body,
    timeout: this.timeout,
    headers: {
      'Content-Length': body.length,
      'Content-Type': 'text/xml',
      'Accept': 'text/xml',
    },
  }, this.requestOptions);

  request(options, function(err, res, body) {
    if (err) {
      return callback(err);
    }

    if (undefined === body) {
      return callback(new Error('invalid body'));
    }

    if (res.statusCode >= 300) {
      var status = HTTP_STATUS_CODES[res.statusCode];

      var explanation = body.match(/<error>([^<]+)/i);
      explanation = (explanation)
        ? ': ' + explanation[1]
        : ': ' + body;

      return callback(new Error(
        'Notification failed: ' + res.statusCode + ' ' + status + explanation
      ));
    }

    // Give me a break, this is legit : )
    var m = body.match(/<url>([^<]+)/i);
    var url = (m)
      ? m[1]
      : null;

    callback(null, url);
  });
}

Airbrake.prototype.notify = function(err, cb) {
  var callback = this._callback(cb);
  // log errors instead of posting to airbrake if a dev enviroment
  if (this.developmentEnvironments.indexOf(this.env) != -1) {
    this.log(err);
    return callback(null, null, true);
  }

  return this._sendRequest(err, callback);
};

Airbrake.prototype.notify = function(err, cb) {
  var callback = this._callback(cb);
  var exit = false
  // log errors instead of posting to airbrake if a dev enviroment
  if (this.developmentEnvironments.indexOf(this.env) != -1) {
    this.log(err);
    return callback(null, null, true);
  }
  this.ignoredExceptions.forEach(function(exception){
    if (err instanceof exception){
      exit = true;
    }
  })

  if (exit){
    return callback(null, null, false);
  }

  return this._sendRequest(err, callback);
};

Airbrake.prototype.notify = function(err, cb) {
  var callback = this._callback(cb);
  var exit = false
  // log errors instead of posting to airbrake if a dev enviroment
  if (this.developmentEnvironments.indexOf(this.env) != -1) {
    this.log(err);
    return callback(null, null, true);
  }
  this.ignoredExceptions.forEach(function(exception){
    if (err instanceof exception){
      exit = true;
    }
  })

  if (exit){
    return callback(null, null, false);
  }

  return this._sendRequest(err, callback);
};

Airbrake.prototype._callback = function(cb) {
  var self = this;
  return function(err) {
    if (cb) {
      cb.apply(self, arguments);
      return;
    }

    if (err) {
      self.emit('error', err);
    }
  };
};

Airbrake.prototype.url = function(path) {
  return this.protocol + '://' + this.serviceHost + path;
};

Airbrake.prototype.notifyXml = function(err, pretty) {
  var notice = xmlbuilder.create().begin('notice', {
    version: '1.0',
    encoding: 'UTF-8'
  });

  this.appendHeaderXml(notice);
  this.appendErrorXml(notice, err);
  this.appendRequestXml(notice, err);
  this.appendServerEnvironmentXml(notice);

  return notice.doc().toString({pretty: pretty});
};

Airbrake.prototype.appendHeaderXml = function(notice) {
  notice
    .att('version', '2.2')
    .ele('api-key')
      .txt(this.key || '-')
    .up()
    .ele('notifier')
      .ele('name')
        .txt(Airbrake.PACKAGE.name)
      .up()
      .ele('version')
        .txt(Airbrake.PACKAGE.version)
      .up()
      .ele('url')
        .txt(Airbrake.PACKAGE.homepage)
      .up()
    .up();
};

Airbrake.prototype.appendErrorXml = function(notice, err) {
  var trace = stackTrace.parse(err);
  var error = notice
    .ele('error')
      .ele('class')
        .txt(err.type || 'Error')
      .up()
      .ele('message')
        .txt(err.message || '-')
      .up()
      .ele('backtrace');

  var self = this;

  trace.forEach(function(callSite) {
    error
      .ele('line')
        .att('method', callSite.getFunctionName() || '')
        .att('file', (callSite.getFileName() || '').replace(self.projectRoot, ""))
        .att('number', callSite.getLineNumber() || '');
  });
};

Airbrake.prototype.appendRequestXml = function(notice, err) {
  var request = notice.ele('request');

  var self = this;
  ['url', 'component', 'action'].forEach(function(nodeName) {
    var node = request.ele(nodeName);
    var val = err[nodeName];

    if (nodeName === 'url') {
      if (!val) {
        val = self.host;
      } else if (val.substr(0, 1) === '/') {
        val = self.host + val;
      }
    }

    if (val) {
      node.txt(val);
    }
  });

  this.addRequestVars(request, 'cgi-data', this.cgiDataVars(err));
  this.addRequestVars(request, 'session', this.sessionVars(err));
  this.addRequestVars(request, 'params', this.paramsVars(err));
};

Airbrake.prototype.addRequestVars = function(request, type, vars) {
  this.emit('vars', type, vars);

  var node;
  Object.keys(vars).forEach(function(key) {
    node = node || request.ele(type);

    var value = vars[key];
    if ('string' !== typeof value) {
      value = util.inspect(value, { showHidden: true, depth: null });
    }

    value = value.replace(/[\u0000-\u0008\u000B-\u000C\u000E-\u001F\uD800-\uDFFF\uFFFE-\uFFFF]/g, " ");

    node
      .ele('var')
      .att('key', key)
      .txt(value);
  });
};

Airbrake.prototype.cgiDataVars = function(err) {
  var cgiData = {};
  Object.keys(process.env).forEach(function(key) {
    cgiData[key] = process.env[key];
  });

  var exclude = [
    'type',
    'message',
    'arguments',
    'stack',
    'url',
    'session',
    'params',
    'component',
    'action',
  ];

  Object.keys(err).forEach(function(key) {
    if (exclude.indexOf(key) >= 0) {
      return;
    }

    cgiData['err.' + key] = err[key];
  });

  cgiData['process.pid'] = process.pid;

  if(os.platform() != "win32") {
    // this two properties are *NIX only
    cgiData['process.uid'] = process.getuid();
    cgiData['process.gid'] = process.getgid();
  }

  cgiData['process.cwd'] = process.cwd();
  cgiData['process.execPath'] = process.execPath;
  cgiData['process.version'] = process.version;
  cgiData['process.argv'] = process.argv;
  cgiData['process.memoryUsage'] = process.memoryUsage();
  cgiData['os.loadavg'] = os.loadavg();
  cgiData['os.uptime'] = os.uptime();

  return cgiData;
};

Airbrake.prototype.sessionVars = function(err) {
  return (err.session instanceof Object)
    ? err.session
    : {};
};

Airbrake.prototype.paramsVars = function(err) {
  return (err.params instanceof Object)
    ? err.params
    : {};
};

Airbrake.prototype.appendServerEnvironmentXml = function(notice) {
  var serverEnvironment = notice.ele('server-environment');

  if (this.projectRoot) {
    serverEnvironment
      .ele('project-root')
      .txt(this.projectRoot);
  }

  serverEnvironment
      .ele('environment-name')
      .txt(this.env);

  if (this.appVersion) {
    serverEnvironment
      .ele('app-version')
      .txt(this.appVersion);
  }
};

Airbrake.prototype.trackDeployment = function(params, cb) {
  if (typeof params === 'function') {
    cb = params;
    params = {};
  }

  params = hashish.merge({
    key: this.key,
    env: this.env,
    user: process.env.USER,
    rev: '',
    repo: '',
  }, params);

  var body = this.deploymentPostData(params);

  var options = hashish.merge({
    method: 'POST',
    url: this.url('/deploys.txt'),
    body: body,
    timeout: this.timeout,
    headers: {
      'Content-Length': body.length,
      'Content-Type': 'application/x-www-form-urlencoded',
    },
  }, this.requestOptions);

  var callback = this._callback(cb);

  request(options, function(err, res, body) {
    if (err) {
      return callback(err);
    }

    if (res.statusCode >= 300) {
      var status = HTTP_STATUS_CODES[res.statusCode];
      return callback(new Error(
        'Deployment failed: ' + res.statusCode + ' ' + status + ': ' + body
      ));
    }

    callback(null, params);
  });
};

Airbrake.prototype.deploymentPostData = function(params) {
  return querystring.stringify({
    'api_key': params.key,
    'deploy[rails_env]': params.env,
    'deploy[local_username]': params.user,
    'deploy[scm_revision]': params.rev,
    'deploy[scm_repository]': params.repo
  });
};
